src/transformStateWithClones.js

```diff
@@ -6,8 +6,71 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const historyArray = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    const newCopy = Object.assign({}, historyArray[historyArray.length - 1]);
+    let stateStep;
+
+    switch (action.type) {
+      case 'addProperties':
+        if (historyArray.length >= 1) {
+          stateStep = addProperties(newCopy, action.extraData);
+          break;
+        } else {
+          stateStep = addProperties(stateCopy, action.extraData);
+          break;
+        }
+
+      case 'removeProperties':
+        if (historyArray.length >= 1) {
+          stateStep = removeProperties(newCopy, action.keysToRemove);
+          break;
+        } else {
+          stateStep = removeProperties(stateCopy, action.keysToRemove);
+          break;
+        }
+
+      case 'clear':
+        if (historyArray.length >= 1) {
+          stateStep = clearProperties(newCopy);
+          break;
+        } else {
+          stateStep = clearProperties(stateCopy);
+          break;
+        }
+
+      default:
+        return 'Uncorrectable types';
+    }
+
+    historyArray.push(stateStep);
```

Comment:
The `historyArray.push(stateStep);` line should be moved outside of the `switch` block, after the current action is processed, to reduce redundancy and improve code clarity. This aligns with the task requirement to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block.